### PR TITLE
feat(federation): introduce private key in community table

### PR DESCRIPTION
**Pull request by ulfgebhardt**
### Issues

- based on https://github.com/gradido/gradido/pull/3023
- replaces https://github.com/gradido/gradido/pull/2970

### Todo

- [X] None

### CODE REVIEW

File: database/entity/0067-private_key_in_community_table/Community.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  @Column({ name: 'foreign', type: 'bool', nullable: false, default: true })
+  foreign: boolean
+
+  @Column({ name: 'url', length: 255, nullable: false })
+  url: string
+
+  @Column({ name: 'public_key', type: 'binary', length: 64, nullable: false })
+  publicKey: Buffer
+
+  @Column({ name: 'private_key', type: 'binary', length: 64, nullable: true })
```
Comment (clauspeterhuebner) [inline database/entity/0067-private_key_in_community_table/Community.ts:24]: why not setting the length to 32 directly for this new column, if you for yourself work on reducing the length for keypairs?

File: database/migrations/0067-private_key_in_community_table.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * This migration adds a field for the private key in the community.table
+ */
+
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+export async function upgrade(queryFn: (query: string, values?: any[]) => Promise<Array<any>>) {
+  await queryFn(
+    'ALTER TABLE `communities` ADD COLUMN `private_key` binary(64) DEFAULT NULL AFTER `public_key`;',
```
Comment (clauspeterhuebner) [inline database/migrations/0067-private_key_in_community_table.ts:11]: question about the length? see above.

File: dht-node/src/dht_node/index.ts [hunk 3]
```diff
@@ -29,14 +24,20 @@ export type CommunityApi = {
   url: string
 }
 
+type KeyPair = { publicKey: Buffer; secretKey: Buffer }
+
 export const startDHT = async (topic: string): Promise<void> => {
   try {
     const TOPIC = DHT.hash(Buffer.from(topic))
-    const keyPair = DHT.keyPair(getSeed())
+    const keyPair = DHT.keyPair(
```
Comment (clauspeterhuebner) [inline dht-node/src/dht_node/index.ts:32]: missing the comment like in other PRs
it's a pain to review several times the same code with the same remarks

**Review by clauspeterhuebner — DISMISSED** [review]

File: database/entity/0067-private_key_in_community_table/Community.ts (hunk 1 above)
Comment (ulfgebhardt) [reply database/entity/0067-private_key_in_community_table/Community.ts:24]: The private key has a length of 64 bytes, the public key has 32 bytes.
This is the private key.

Comment (ulfgebhardt) [reply database/entity/0067-private_key_in_community_table/Community.ts:24]: This PR handles to correct the public_key size: https://github.com/gradido/gradido/pull/3026
It is unreleated to this PR

Comment (clauspeterhuebner) [reply database/entity/0067-private_key_in_community_table/Community.ts:24]: Ok, i understand...

**Review by clauspeterhuebner — APPROVED** [review]

**Pull request merged**